### PR TITLE
feat: added diagonal frontend

### LINKED ISSUE #26377: diagonal
Paddle Diagonal

### PR DESCRIPTION
**Pull request by lukaM19**
# PR Description
feat: added diagonal frontend

## Related Issue

Close #26377

## Checklist

- [x] Did you add a function?
- [x] Did you add the tests?
- [x] Did you run your tests and are your tests passing?
- [x] Did pre-commit not fail on any check?
- [x] Did you follow the steps we provided?

### CODE REVIEW

**Comment by lukaM19** [comment]
Doesn't seem to cause any new test failures ! Let me know if I can edit something!

**Comment by lukaM19** [comment]
@KareemMAX , Hi! I just marked my PR ready for review as it doesn't seem to introduce any new failures, if you think I should change something please lmk and I'll get it fixed asap.

File: ivy_tests/test_ivy/test_frontends/test_paddle/test_linalg.py [hunk 1]
```diff
@@ -688,6 +699,49 @@ def test_paddle_eigvalsh(
     )
 
 
+# diagonal
+@handle_frontend_test(
+    fn_tree="paddle.linalg.diagonal",
+    dtype_and_values=helpers.dtype_and_values(
+        available_dtypes=helpers.get_dtypes("float"),
```
Comment (KareemMAX) [inline ivy_tests/test_ivy/test_frontends/test_paddle/test_linalg.py:706]: I guess this should be changed to `valid`, you may do the necessary changes as well

**Review by KareemMAX — CHANGES_REQUESTED** [review]
Thank you so much @lukaM19,
LGTM as for now, I've just a small change. Once done we can run the tests

**Comment by lukaM19** [comment]
I did the fix you suggested but it introduces a lot of errors with data types, could you suggest a fix

**Comment by KareemMAX** [comment]
> I did the fix you suggested but it introduces a lot of errors with data types, could you suggest a fix, @KareemMAX

You may use `@with_supported_dtypes` as shown [here](https://github.com/unifyai/ivy/pull/26407/files#diff-19d3f3922bf2d8d7df9a9bdd6331822707caeb37e062644d129e8d95454801baR53)

**Comment by lukaM19** [comment]
> > I did the fix you suggested but it introduces a lot of errors with data types, could you suggest a fix, @KareemMAX
> 
> @lukaM19 You may use `@with_supported_dtypes` as shown [here](https://github.com/unifyai/ivy/pull/26407/files#diff-19d3f3922bf2d8d7df9a9bdd6331822707caeb37e062644d129e8d95454801baR53)

@KareemMAX   I tried using that in the last commit but now even more tests fail, to be honest that confused me a little. If you have any ideas as to why I would appreciate any tips!

**Comment by KareemMAX** [comment]
> @KareemMAX I tried using that in the last commit but now even more tests fail, to be honest that confused me a little. If you have any ideas as to why I would appreciate any tips!

@lukaM19, that shouldn't happen. Have you tried running the tests locally?

**Comment by lukaM19** [comment]
@KareemMAX Hi, I tried to fix it and it doesn't seem to fail any tests, let me know if you think I should change something else as well. Best , Luka

**Review by KareemMAX — CHANGES_REQUESTED** [review]
Hey @lukaM19,
Sadly when I run this test `ivy_tests/test_ivy/test_frontends/test_paddle/test_linalg.py::test_paddle_linalg_diagonal` it fails sadly. Can you fix that?

You can run it locally by using this command:
```bash
pytest ivy_tests/test_ivy/test_frontends/test_paddle/test_linalg.py::test_paddle_linalg_diagonal
```

**Comment by lukaM19** [comment]
@KareemMAX  Hi, I removed a redundant test since I realized the same function was being tested twice in test_tensor and test_linalg, still the tests here don't show any errors. I am having trouble running the tests locally even in a docker container, is it possible for you to check again? if you are getting an error what kind of an error is it?

**Comment by KareemMAX** [comment]
> @KareemMAX Hi, I removed a redundant test since I realized the same function was being tested twice in test_tensor and test_linalg, still the tests here don't show any errors.

Actually that test you removed wasn't redundant, and it failing is considered a bug and a blocker to merge the PR. Can you please add it back and fix it?



> I am having trouble running the tests locally even in a docker container, is it possible for you to check again? if you are getting an error what kind of an error is it?

The docker image suppose to work without any additional configuration, can you post the error message you see? And are you sure you are using `unifyai/ivy` image from docker hub?

I can suggest you somethings to do to make sure you are working with the right environment. You can either:
1. Use codespaces, you have free number of hours every month.
2. Use vscode dev containers. It should give you an identical environment to code spaces locally through docker.

Those are the methods I use personally, of course if you are using PyCharm it should work fine when you use `unifyai/ivy` image from docker hub. Here are the instructions for all of the methods for [setting up your environment](https://unify.ai/docs/ivy/overview/contributing/setting_up.html).

**Comment by lukaM19** [comment]
> Hey @lukaM19, Sadly when I run this test `ivy_tests/test_ivy/test_frontends/test_paddle/test_linalg.py::test_paddle_linalg_diagonal` it fails sadly. Can you fix that?
> 
> You can run it locally by using this command:
> 
> ```shell
> pytest ivy_tests/test_ivy/test_frontends/test_paddle/test_linalg.py::test_paddle_linalg_diagonal
> ```

What error message are you getting? I tried running it on codespace and I get the following error messages ` ERROR: not found: /workspaces/ivy/ivy_tests/test_ivy/test_frontends/test_paddle/test_linalg.py::test_paddle_diagonal
(no name '/workspaces/ivy/ivy_tests/test_ivy/test_frontends/test_paddle/test_linalg.py::test_paddle_diagonal' in any of [<Module test_linalg.py>]) ` , in fact I tried running the command for other tests like `test_paddle_matmul` in codespaces and I got the same type of error for all the functions in the file, so I think it might be an issue with the file in general.

But when I test the test_tensor.py test, it fails 3/5 tests with the following error message:`  
  def _get_fn_dtypes(framework: str, kind="valid", mixed_fn_dtypes="compositional"):
        all_devices_dtypes = test_globals.CURRENT_RUNNING_TEST.supported_device_dtypes[
            framework
        ]
        if mixed_fn_dtypes in all_devices_dtypes:
            all_devices_dtypes = all_devices_dtypes[mixed_fn_dtypes]
>       return all_devices_dtypes[test_globals.CURRENT_DEVICE_STRIPPED][kind]
E       KeyError: 'cpu'`
I don't understand what could cause this do you have a suggestion on what to look for?

I only removed the `redundant` test because both tests test the same exact function, there is no difference in calling them so I thought it would be redundant to test the same function, is that wrong?

**Comment by KareemMAX** [comment]
> I only removed the `redundant` test because both tests test the same exact function, there is no difference in calling them so I thought it would be redundant to test the same function, is that wrong?

Sadly they do in fact test different function signatures, this would be important for us later on to find bugs if we changed a specific method for any reason.



> What error message are you getting? I tried running it on codespace and I get the following error messages `ERROR: not found: /workspaces/ivy/ivy_tests/test_ivy/test_frontends/test_paddle/test_linalg.py::test_paddle_diagonal (no name '/workspaces/ivy/ivy_tests/test_ivy/test_frontends/test_paddle/test_linalg.py::test_paddle_diagonal' in any of [<Module test_linalg.py>])` , in fact I tried running the command for other tests like `test_paddle_matmul` in codespaces and I got the same type of error for all the functions in the file, so I think it might be an issue with the file in general.

This seems to be an issue in your setup (most likely). I'll run the tests and provide you with the results, but you can consult `#ivy-tests` at anytime

**Comment by lukaM19** [comment]
How could I edit my setup if I'm running the tests on browser in codespaces?

**Review by KareemMAX — APPROVED** [review]
I looked into the issue and seems like you referenced the wrong function in your test, along with other issues with supported dtypes. I fixed those

**Pull request merged**